MCOL-5559 This patch extends the scope of the crit section to cover remap operation to avoid access races running EMIndex operations

drrtuy opened the pull request:
Similar to MCOL-5487. Remap op wasn't covered by an exclusive shmem lock.

Commit:
fix(extent-map): MCOL-5559 This patch extends the scope of the crit section to cover remap operation to avoid access races running EMIndex operations

tntnatbry commented on versioning/BRM/extentmap.cpp:
```diff
@@ -2080,10 +2092,12 @@ void ExtentMap::grabEMIndex(OPS op)
+                                                     isReadOnly, emIndexLocked, op == READ ? &fMST : nullptr);
 
       if (r_only)
         fPExtMapIndexImpl_->makeReadOnly();
```
`makeReadOnly()` should be under WRITE lock.

drrtuy replied:
JFYI you are mixing locks up. mutex lock that is taken for the scope of the function controls multithread races whilst shmem RWLock controls interprocess races.  fPExtMapIndexImpl_ is process-specific entity.

tntnatbry replied:
`makeReadOnly()` is calling `BRMManagedShmImpl::remap()` under the hood, which is a write op to the shmem segment.

drrtuy replied:
After some consideration I agree with the point.

tntnatbry commented on versioning/BRM/extentmap.cpp:
```diff
@@ -2092,8 +2106,10 @@ void ExtentMap::grabEMIndex(OPS op)
   else if (fPExtMapIndexImpl_->getShmemImplSize() != (unsigned)fEMIndexShminfo->allocdSize)
   {
     fPExtMapIndexImpl_->refreshShm();
```
`refreshShm()` should be under WRITE lock.

drrtuy replied:
Same as previously.

tntnatbry replied:
The write op in `refreshShm()` is not synchronised. Here is the race scenario:
  Reader1: Execute `ExtentMap::getExtents()` -> `fPExtMapIndexImpl_->find()`
  Reader2: Execute `ExtentMap::getExtents()` -> `grabEMEntryTable(READ)` -> `fPExtMapIndexImpl_->refreshShm()`

drrtuy replied:
After some consideration I agree with the point.

tntnatbry reviewed (changes requested):
It looks like similar race also exists in `ExtentMap::grabFreeList()`.

Also, I noticed there is no call to `ExtentMapIndexImpl::refreshShm()` in `DBRM::refreshShm()`.

drrtuy commented:
> It looks like similar race also exists in `ExtentMap::grabFreeList()`.
> 
> Also, I noticed there is no call to `ExtentMapIndexImpl::refreshShm()` in `DBRM::refreshShm()`.

The story is different for FreeList. It is based on a different kind of shmem wrapper, namely unmanaged shmem. JFYI there were no races before we switched EM and introduced EMIndex on top of managed shmem implementation. Thus I won't change it in 23.02 but I have the fact in mind doing the patch for 23.08.
